pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -166,6 +192,11 @@ func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCl
 		volumeIndex.Add(volume)
 	}
 
+	if isBeingProvisioned(volume) {
+		glog.V(4).Infof("Skipping PersistentVolume[%s], waiting for provisioning to finish", volume.Name)
+		return nil
+	}
+
 	switch currentPhase {
```

Comment:
The binder should ignore the claim if it's not provisioned.  The
provisioner makes a PV w/ ClaimRef, which makes it bound. Only one
controller should be making state changes at a time. At least, that's the
idea. Only 1 ever sets "completed", for example.

I will add an integration test with many volumes and claims and hammer the
controllers.

On Thursday, December 10, 2015, Saad Ali <notifications@github.com> wrote:

> In
> pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go
> <https://github.com/kubernetes/kubernetes/pull/14537#discussion_r47315024>
> :
>
> > @@ -166,6 +192,11 @@ func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCl
> >  		volumeIndex.Add(volume)
> >  	}
> >
> > +	if isBeingProvisioned(volume) {
> > +		glog.V(4).Infof("Skipping PersistentVolume[%s], waiting for provisioning to finish", volume.Name)
> > +		return nil
> > +	}
> > +
> >  	switch currentPhase {
>
> It feels like there will be race conditions between the binder and
> provisioner controllers since both operate on the same API objects without
> any apparent locking mechanism before them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14537/files#r47315024>.
>